### PR TITLE
Support connections to vendor-specific interface classes

### PR DESCRIPTION
**Pull request by reillyeon**
This change refactors the polyfill so that it can be configured to filter or and connect to vendor-specific interface classes rather than the standard USB CDC-ACM class.

### CODE REVIEW

File: serial.ts [hunk 1]
```diff
@@ -53,119 +81,291 @@ const kParityIndexMapping: ParityType[] =
...
+          }));
+      this.writable = new WritableStream(
+          new UsbEndpointUnderlyingSink(this.device_, this.outEndpoint_),
+          new ByteLengthQueuingStrategy({
+            highWaterMark: this.serialOptions_.buffersize,
+          }));
     } catch (error) {
+      if (this.device_.opened) {
+        await this.device_.close();
+      }
```
Comment (jameshollyergoogle) [inline serial.ts]: Seems like we should set dtr back to false if we don't have the streams working.  Maybe just call close()?
Not sure how readable.cancel() this.writable.abort() calls would behave if you called close now though.

File: serial.ts [hunk 2]
```diff
@@ -53,119 +81,291 @@ const kParityIndexMapping: ParityType[] =
...
    */
-  private setPort(device: USBDevice) {
-    if (!SerialPort.isSerialDevice(device)) {
-      throw new TypeError('This is not a serial port');
+  public async setSignals(signals: SerialOutputSignals): Promise<void> {
+    this.outputSignals_ = {...this.outputSignals_, ...signals};
+
+    if (signals.dtr !== undefined || signals.rts !== undefined) {
+      const value = (this.outputSignals_.dtr ? 1 << 0 : 0) |
+                    (this.outputSignals_.rts ? 1 << 1 : 0);
```
Comment (jameshollyergoogle) [inline serial.ts]: I think this calculation deserves a bit of explanation/background.  Maybe just a link to s spec or something?

File: serial.ts [hunk 3]
```diff
@@ -53,119 +81,291 @@ const kParityIndexMapping: ParityType[] =
...
+ * Utility function to get the interface implementing a desired class.
+ * @param {USBDevice} device The USB device.
+ * @param {number} classCode The desired interface class.
+ * @return {USBInterface} The first interface found that implements the desired
+ * class.
+ * @throws TypeError if no interface is found.
+ */
+function findInterface(device: USBDevice, classCode: number): USBInterface {
+  const configuration = device.configurations[0];
+  for (const iface of configuration.interfaces) {
```
[the remark this answers is not in the document]
Comment (reillyeon) [reply serial.ts]: `interface` is a reserved word.

File: serial.ts [hunk 4]
```diff
@@ -53,119 +81,291 @@ const kParityIndexMapping: ParityType[] =
...
+          controller.enqueue(result.data);
+        },
+        (error) => {
+          controller.error(error.toString());
+        });
+  }
+}
+
+/**
+ * Implementation of UnderlyingSink for USB devices.
```
[the remark this answers is not in the document]
Comment (reillyeon) [reply serial.ts]: I've added a link to the WHATWG Streams specification.

File: serial.ts (hunk 2 above)
Comment (reillyeon) [reply serial.ts]: I've added links to the specification.

File: serial.ts (hunk 1 above)
Comment (reillyeon) [reply serial.ts]: Construction of the streams won't fail as currently implemented. I've updated the steps for `close()` however to consider the case where the streams have not yet been created.

File: serial.ts [hunk 5]
```diff
@@ -295,20 +335,20 @@ export class SerialPort {
...
+    if (this.writable_) {
+      promises.push(this.writable_.abort());
     }
     await Promise.all(promises);
-    await this.setSignals({dtr: false});
-    await this.device_.close();
+    this.readable_ = null;
+    this.writable_ = null;
+    if (this.device_.opened) {
+      await this.setSignals({dtr: false});
```
Comment (jameshollyergoogle) [inline serial.ts]: I would think that is we are closing this port then we are also not ready to send either.  Should we also set RTS to false?

Comment (reillyeon) [reply serial.ts]: Done.

**Pull request merged**